README.md

```diff
@@ -76,32 +76,59 @@ For instance, `wnfs/APPEND` is an action for WebNative filesystem paths. The act
 
 A capability is the association of an action to a resource: `resource x action`
 
-## 2.4 Scope
-
-An authorization scope is a set of capabilities. Scopes MUST compose with set semantics, so multiple scopes in an array MAY be considered the (deduplicated) union of all of the inner scopes.
-
-![](./assets/scope_union.jpg)
+## 2.4 Capability Pool
+
+A capability pool is a set of capabilities. Pools MUST compose with set semantics, so multiple pools in an array MAY be considered the (deduplicated) union of all of the inner capabilities.
+
+``` plaintext
+                 ┌───────────────────┐  ─┐
+                 │                   │   │
+                 │                   │   │
+┌────────────────┼───┐               │   │
+│                │   │ Resource B    │   │     
+│                │   │               │   │       BxZ
+│                │   │     X         │   ├─── Capability
+│     Resource A │   │               │   │
+│                │   │ Ability Z     │   │
+│         X      │   │               │   │
+│                │   │               │   │
+│     Ability Y  │   │               │   │
+│                └───┼───────────────┘  ─┘
+│                    │
+│                    │
+└────────────────────┘
+
+└──────────────────┬─────────────────┘
+                   │
+
+               AxY U BxZ
+                  Pool
+```
 
-The "scope" is the total rights of the authorization space down to the relevant volume of authorizations. With the exception of [rights amplification](#54-rights-amplification), every individual delegation MUST have equal or less scope from their delegator. Inside this content space, you can draw a boundary around some resource(s) (their type, identifiers, and paths or children), and their capabilities.
+The "pool" is the total rights of the authorization space down to the relevant volume of authorizations. Individual capabilities MAY overlap; the pool is the union. With the exception of [rights amplification](#54-rights-amplification), every individual delegation MUST have equal or narrower capabilties from their delegator. Inside this content space, you can draw a boundary around some resource(s) (their type, identifiers, and paths or children), and their capabilities.
 
-As a practical matter, since scopes form a group, you can be fairly loose: order doesn’t matter, and merging resources can be quite broad since the more capable of any overlap will take precedence (i.e. you don’t need a clean separation).
+As a practical matter, since pools form a group, you can be fairly loose: order doesn’t matter, and merging resources can be quite broad since the more capable of any overlap will take precedence (i.e. you don’t need a clean separation).
```

Comment:
@QuinnWilton

_This_ I could use a hand with while WIP, though! The feedback is that "scope" is confusing in a programming context (we use that word a LOT). It does really capture the idea, and I'm struggling to find a replacement that really fits.

It could be "capability set", but that doesn't feel very descriptive. "Pool" came up because I had a side conversation with Philipp about worker pools, and it's amorphous enough, but still doesn't really capture the intent.

Any thoughts?

Reply:
🤔 "Collection"

Reply:
`<mumbles something about the wallet metaphor being terrible>`

Reply:
How do you feel about "capability space"? I've been visualizing the capabilities carried by a UCAN as forming a 3D space with resources, actions, and identities along each axis, which would mean that the capabilities represented by the union of two delegations would be given by the union of their capability spaces.

Reply:
I suppose you already use the term "authorization space" here, so you already get into that idea a bit, but I am a big fan of the geometric interpretation, so something along those lines might be helpful: "capability boundary", "capability sphere", etc

Reply:
"capability sphere" doesn't quite feel right, because that implies a symmetry that doesn't necessarily exist. I think "capability boundary" is really the closest name for what I'm visualizing, if you think of it from a topological sense.

Reply:
Alright, I think I've traced some of my confusion to the use of "paths or children" in the above passage. I'm guessing that's referring specifically to Webnative, so I'm going to make some assumptions here to clarify my understanding of things.

It's possible that all of this is entirely wrong though, because my interpretation of this spec is mostly based on this document and some educated guesses. If not, I'm no longer convinced that a geometric interpretation is helpful, and I'll to explain why by considering WebNative paths and email addresses. I'll assume that paths form a tree, and that a capability against some path also applies to all of that path's descendants.

I'll also assume that we have four actions:
```
wnfs/APPEND
wnfs/OVERWRITE (implies wnfs/APPEND)
wnfs/READ
email/SEND
```

Some filesystem:
```
Private/
├─ Pictures/
├─ Documents/
```

Some email addresses:
```
alice@example.com
bob@example.com
```

And two capabilities:

```
A = wnfs/APPEND × "Private/"
B = wnfs/OVERWRITE × "Private/Documents/"
```

Then:
- `A` grants `wnfs/APPEND` against the entire filesystem
- `B` grants `wnfs/APPEND` and `wnfs/OVERWRITE` against `"Private/Documents/"`
- `wnfs/READ` has not been granted against any paths
- `email/SEND` has not been granted against any email addresses

The spec already mentions that actions can take on any semantics, but this example shows that the same is true for resources. The paths form a partial order, over the "descendant of" relationship, but there's no way to compare paths that aren't descendants. Similarly, `{ wnfs/APPEND, wnfs/OVERWRITE }` forms a partial order, but there's no relationship between email addresses whatsoever, email addresses and paths, and `wnfs/READ` or `email/SEND` with any other actions.

All of this together means that we don't have one connected space to represent all of the possible capabilities in, and instead have a disconnected space made up of three subspaces:
1) Write capabilities against filesystem paths, with points in `{ NONE, APPEND, OVERWRITE }`
2) Read capabilities against filesystem paths, with points in `{ NONE, READ }`
3) Send capabilities against email addresses, with points in `{ NONE, SEND }`

If I'm understanding your description of things correctly, then what you're tentatively calling a "capability pool" refers to one of these subspaces. Which is to say that there is a way to make a geometric / topological interpretation of this stuff work, but I think it's too confusing to be elucidating, and requires far too much knowledge of the underlying semantics of the resources and the actions to easily visualize.

Maybe this can give us a starting point for thinking about clearer language though? At the very least, this leaves me feeling like what you're against calling a "capability set" is likely more accurately described as a "capability poset", though I don't think that's a great term either.

My favourite term is still "capability boundary", but I'm wary of the complexity involved in conveying the motivations behind the term. That being said, since posets can be viewed as trees, maybe the problem can shifted toward one of viewing the authorization space as a forest of capabilities?

Then a "capability pool" would become a "capability tree", and would compose under the semantics given by the underlying `(resource, action)` pair. My main reservation here comes from not knowing whether that structure is guaranteed to be a poset, or if richer structures like lattices are possible. I'm assuming not, because all of this is being built on top of IPFS, but I want to raise the concern!

The tree interpretation is still pretty abstract, but maybe with enough concrete examples it would be a more natural perspective on the problem, given that graphs are likely more familiar to most engineers than topological spaces?

Anyway, I have no idea if this is all gibberish or not, but I'm happy to chat with you in more detail about it either way!

Reply:
> How do you feel about "capability space"?

I tend to use this as "the complete set of all possible capabilities", in the same sense as "the solution space". But maybe that's just me?

Reply:
>  but this example shows that the same is true for resources

Yup! They're literally URIs, so however that URI works is their semantics. Perhaps good to spell that out explicitly 🤔 👍

Reply:
>  I worry that it's too confusing to be elucidating, and requires too much knowledge of the underlying semantics of the resources and the actions to easily visualize.

Oh, you're going way deeper! I want a term that captures the idea of the union of all capabilities that are currently available to the current agent (whoever has its DID) that are provided by this UCAN. It's a set union on all of their capabilities, but I don't want to use the phrase "set union over capabilities" because it's overly technical and wordy.

The other thing that I'm trying to capture is the concept of overlapping. For example, we could have the following:

```
"wnfs:/.../pictures" x WRITE
"wnfs:/.../pictures/vacation" x DESTROY
```

Where DESTROY > WRITE. These overlap, but not perfectly (to say nothing of overlapping paths in a DAG).

What I liked about "scope" was that it's essentially your Γ. Everything that's in your execution environment with the UCAN loaded, which is to say "the current scope".

Reply:
> not knowing whether those semantics are guaranteed to be a poset

They are! We use posets in the code during validation.

_Each_ capability forms a DAG of capability "strength" or "size" (to use the visual metaphor). Technically having several is like an ad hoc version of that same idea (literally an algebraic product type). But! I don't want to scare people off with lots of math.

Reply:
(You see how I'm struggling here? 😅😅😅 Why is naming things so hard?!)

Reply:
My opinion is that scope is the right word. Maybe changing that section title to “Scope of Authority” would help to introduce the type of scope being discussed.

I looked at a couple of synonyms like “purview” or “ambit”, but those don’t make sense in a programming context. 😆

Reply:
> > How do you feel about "capability space"?
> 
> I tend to use this as "the complete set of all possible capabilities", in the same sense as "the solution space". But maybe that's just me?

I agree with this!

> > but this example shows that the same is true for resources
> 
> Yup! They're literally URIs, so however that URI works is their semantics. Perhaps good to spell that out explicitly 🤔 👍

I would lean toward being explicit, but I don't think you need to get too deep about it. Just a quick sentence like "Each resource MAY have its own semantics. For example, WebNative filesystem paths form a hierarchy, and can be visualized as a tree." would make explicit that the paths can behave more like a tree than like opaque identifiers.

> What I liked about "scope" was that it's essentially your Γ. Everything that's in your execution environment with the UCAN loaded, which is to say "the current scope". You don't really need to think about the interaction so much, since they're all phrased in the positive, "any witness will do"

This is a really good insight, and maybe the way to express this is just by connecting back to section 1.3: "Each UCAN includes a constructive set of assertions of what it is allowed to do. Note that this is not a predicate: it is a positive assertion of rights."

Having thought through this conversation, I think I agree with you and @bgins about scope being a really good word to use here, so maybe the terms themselves don't actually need to change, and you just need to define them upfront and give an example? Something like:

> The set of capabilities granted by a UCAN is called its scope. Scopes give a declarative description of the granted abilities, and two scopes can be combined by concatenating all of their capabilities together. Since broader capabilities absorb weaker capabilities, this process is additive, and scopes can be combined in any order, with the result always being at least as broad as each of the original scopes.
>
> For example, given the following scopes against a WebNative filesystem, they can be merged as follows:
> ```
> ScopeA = { "wnfs:/.../pictures" x WRITE }
> ScopeB = { "wnfs:/.../pictures/vacation" x DESTROY } 
> 
> merge(ScopeA, ScopeA) => {
>    "wnfs:/.../pictures" x WRITE,
>    "wnfs:/.../pictures/vacation" x DESTROY
> }
> ```
> This new scope partially overlaps, and provides both `WRITE` and `DESTROY` capabilities against `"wnfs:/.../pictures/vacation"`. 

Do you think a description like that might address any of the confusion?

> They are! We use posets in the code during validation.

Awesome, that's good to know! I agree about not wanting to overload readers with any of the theory, but that's very helpful for my own understanding of things. Thanks for bearing with me while I wrapped my head around things!

Reply:
> (You see how I'm struggling here? 😅😅😅 Why is naming things so hard?!)

Something something names are the limit of our world something something

Reply:
If you don't want to use the word "scope", then you might be able to substitute "domain"? I'm reminded of realms in [RFC 2617](https://datatracker.ietf.org/doc/html/rfc2617#section-1.2):

> The realm value, in combination with the canonical root URL of the server being accessed, defines the protection > space.
> ...
> The protection space determines the domain over which credentials can
> be automatically applied.

It's not a perfect analogy, but there may be something to steal from there?

Reply:
> > The set of capabilities granted by a UCAN is called its scope. Scopes give a declarative description of the granted abilities, and two scopes can be combined by concatenating all of their capabilities together. Since broader capabilities absorb weaker capabilities, this process is additive, and scopes can be combined in any order, with the result always being at least as broad as each of the original scopes.
> > For example, given the following scopes against a WebNative filesystem, they can be merged as follows:
> > ```
> > ScopeA = { "wnfs:/.../pictures" x WRITE }
> > ScopeB = { "wnfs:/.../pictures/vacation" x DESTROY } 
> > 
> > merge(ScopeA, ScopeA) => {
> >    "wnfs:/.../pictures" x WRITE,
> >    "wnfs:/.../pictures/vacation" x DESTROY
> > }
> > ```
> > This new scope partially overlaps, and provides both `WRITE` and `DESTROY` capabilities against `"wnfs:/.../pictures/vacation"`.

This explanation using scope reads great to me! :clap: